src/main/kotlin/com/fasterxml/jackson/module/kotlin/Extensions.kt

```diff
@@ -50,21 +51,58 @@ fun ObjectMapper.registerKotlinModule(initializer: KotlinModule.Builder.() -> Un
 
 inline fun <reified T> jacksonTypeRef(): TypeReference<T> = object: TypeReference<T>() {}
 
+/**
+ * It is public due to Kotlin restrictions, but should not be used externally.
+ */
+inline fun <reified T> Any?.checkTypeMismatch(): T {
+    // Basically, this check assumes that T is non-null and the value is null.
+    // Since this can be caused by both input or ObjectMapper implementation errors,
+    // a more abstract RuntimeJsonMappingException is thrown.
+    if (this !is T) {
+        // Since the databind implementation of MappingIterator throws RuntimeJsonMappingException,
+        // JsonMappingException was not used to unify the behavior.
+        throw RuntimeJsonMappingException(
+            "Deserialized value did not match the specified type; " +
+                    "specified ${T::class.qualifiedName} but was ${this?.let { it::class.qualifiedName }}"
```

Comment:
Is the error message easy to understand?
For example, if a non-null `String` is specified but `null` is read, the message will look like this

```
Deserialized value did not match the specified type; specified kotlin.String but was null
```

Reply:
Since this is about `null` and `non-null`, including those two words (null and non-null) might be easier to grasp context at first sight?

Reply:
It would be good to  also add (non-null vs null) like

```
Deserialized value did not match the specified type; specified kotlin.String (non-null) but was null
```

Reply:
I made the following

**non-null**

```
Deserialized value did not match the specified type; specified kotlin.String(non-null) but was null
```

**nulable**

```
Deserialized value did not match the specified type; specified kotlin.String? but was kotlin.Int
```